File: leetcode/medium/3499-maximize-active-section-with-trade-i/solution.py
class Solution:
    def maxActiveSectionsAfterTrade(self, s: str) -> int:
        ones = s.count('1')

        t = '1' + s + '1'

        ch = []
        length = []

        i = 0
        while i < len(t):
            j = i
            while j < len(t) and t[j] == t[i]:
                j += 1
            ch.append(t[i])
            length.append(j - i)
            i = j

        maxGain = 0

        for k in range(1, len(ch) - 1):
            if ch[k] == '1' and ch[k-1] == '0' and ch[k+1] == '0':
                maxGain = max(maxGain, length[k-1] + length[k+1])

        return ones + maxGain

File: leetcode/medium/3499-maximize-active-section-with-trade-i/solution.java
class Solution {
    public int maxActiveSectionsAfterTrade(String s) {
        int ones = 0;
        for (char c : s.toCharArray())
            if (c == '1') ones++;

        String t = "1" + s + "1";
        int n = t.length();

        int maxGain = 0;
        int i = 0;

        while (i < n) {
            char prevChar = t.charAt(i);
            int prevLen = 0;
            while (i < n && t.charAt(i) == prevChar) {
                prevLen++;
                i++;
            }

            if (i >= n) break;

            char midChar = t.charAt(i);
            int midLen = 0;
            while (i < n && t.charAt(i) == midChar) {
                midLen++;
                i++;
            }

            if (i >= n) break;

            char nextChar = t.charAt(i);
            int nextLen = 0;
            while (i < n && t.charAt(i) == nextChar) {
                nextLen++;
                i++;
            }

            if (prevChar == '0' && midChar == '1' && nextChar == '0') {
                maxGain = Math.max(maxGain, prevLen + nextLen);
            }

            i -= nextLen;
        }

        return ones + maxGain;
    }
}